cloverage/src/cloverage/source.clj

```diff
@@ -1,27 +1,46 @@
 (ns cloverage.source
-  (:require [clojure.tools.reader :as r]
-            [clojure.tools.reader.reader-types :as rt]
-            [clojure.java.io :as io]
-            [clojure.string :as str])
-  (:import (java.io InputStreamReader File PushbackReader)
-           (clojure.lang RT)))
+  (:require [clojure.java.io :as io]
+            [clojure.string :as str]
+            [clojure.tools.namespace.file :as ns.file]
+            [clojure.tools.reader :as r]
+            [clojure.tools.reader.reader-types :as rt])
+  (:import clojure.lang.RT
+           [java.io File InputStreamReader PushbackReader]))
 
 (defn- get-loader ^ClassLoader []
   (RT/baseLoader))
 
-(defn- resource-exists? [path]
-  (not (nil? (io/resource path (get-loader)))))
-
 (defn resource-path
-  "Given a symbol representing a lib, return a classpath-relative path.  Borrowed from core.clj."
-  [lib]
-  (let [base (.. (name lib)
+  "Given a symbol naming a namespace, return a classpath-relative path to the file defining that namespace.
+
+    (resource-path 'cloverage.source)
+    ;; -> \"cloverage/source.clj\""
+  [ns-symbol]
+  (let [base (.. (name ns-symbol)
                  (replace \- \_)
                  (replace \. \/))]
-    (->> ["clj" "cljc"]
-         (map #(str base "." %))
-         (filter resource-exists?)
-         first)))
+    (some (fn [extension]
+            (let [filename (str base extension)]
+              ;; io/resource will return nil if the file doesn't exist
+              (when (io/resource filename (get-loader))
+                filename)))
+          ns.file/clojure-extensions)))
+
+(defn ns-form
+  "Return the `ns` declaration form from the file associated with the namespace named by `ns-symbol`.
+
+    (ns-form 'cloverage.source)
+    ;; ->
+    (ns cloverage.source
+      (:require [clojure.java.io :as io]
+                [clojure.string :as str]
+                [clojure.tools.namespace.file :as ns.file])
+      (:import clojure.lang.RT
+               java.io.File))"
+  [ns-symbol]
+  (some->> (resource-path ns-symbol)
+           (.getResource (get-loader))
+           ns.file/read-file-ns-decl))
```

Comment:
I tweaked this code to use the `tools.namespace` code for reading the `ns` declaration which seemed safer than relying on our own implementation.